polytracker/include/taintdag/outputfile.hpp

```diff
@@ -0,0 +1,159 @@
+// into a single entry.
+template <typename T> struct SectionArg {
+  T &output_file;
+  std::span<uint8_t> range;
+};
+
+// Container format for storing information about instrumented run.
+//
+// An OutputFile is made up from Sections. Each section is assigned
+// a range of memory it can use freely. Requirements on a section
+// is captured in the Section concept.
+template <Section... Sections> class OutputFile {
+
+  static_assert(sizeof...(Sections) > 0,
+                "OutputFile requires at least one section");
+
+public:
+  // Represents information about a section in the file
+  struct SectionMeta {
+    // Unique tag for this section, identifies the type of section
+    uint32_t tag{0};
+
+    // Alignment of entries in this section
+    uint32_t align{1};
+
+    // Offset from start of file to section start
+    uint64_t offset{0};
+
+    // Size, in bytes, of the section
+    uint64_t size{0};
+  };
+
+  // File header information, followed by section_count SectionMeta.
+  struct FileMeta {
+    char tdag[4]{'T', 'D', 'A', 'G'};
+
+    // Magic version number. Computed from a combination of all section tags.
+    // TODO(hbrodin): There wasn't any real thought process/evaluation of the
+    // distribution of magic numbers involved when coming up with this. Also, it
+    // could probably be removed because all information is available in the
+    // sections.
+    uint16_t magic{(Sections::tag + ...) ^ sizeof...(Sections)};
+
+    // Number of sections in this TDAG
+    uint16_t section_count{sizeof...(Sections)};
+  };
+
+  struct FileHeader {
+    FileMeta meta;
+    SectionMeta sections[sizeof...(Sections)] = {
+        (SectionMeta{.tag = Sections::tag,
+                     .align = Sections::align_of,
+                     .offset = 0,
+                     .size = Sections::allocation_size})...};
+  };
+
+  OutputFile(std::filesystem::path const &filename)
+      : mm_{std::move(filename), required_allocation_size()},
+        hdr_{new (mm_.begin) FileHeader}, alloc_ptr_{mm_.begin +
+                                                     sizeof(FileHeader)},
+        sections_{(SectionArg<OutputFile>{
+            .output_file = *this, .range = do_allocation<Sections>()})...} {
+    // Assumes that the mmap:ed memory is page aligned and FileHeader
+    // has less alignment requirements.
+    if (reinterpret_cast<uintptr_t>(mm_.begin) % alignof(FileHeader) != 0) {
+      error_exit("Mapped memory does not meet alignment requirement of FileHeader");
+    }
+  }
+
+  ~OutputFile() {
+    // Update the memory actually used by each section
+    // TODO(hbrodin): Consider other implementation strategies.
+    size_t _[] = {
+        hdr_->sections[util::TypeIndex<Sections,
+                                       std::tuple<Sections...>>::index]
+            .size = std::get<Sections>(sections_).size()...};
+
+    (void)_;
+    // TODO(hbrodin): Is there a need to notify the sections about shutdown in
```

Comment:
What I'm thinking is that we might want to do per section functionality during shutdown/destruction. It can be done in the section destructors, but those are run after this destructor is run. So, anything we want to have influencing the contents of the tdag file would have to happen before.